ecs-agent/acs/session/session.go

```diff
@@ -11,13 +11,348 @@
 // express or implied. See the License for the specific language governing
 // permissions and limitations under the License.
 
+// Package session deals with appropriately reacting to all ACS messages as well
+// as maintaining the connection to ACS.
 package session
 
 import (
+	"context"
+	"io"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/ecs-agent/api"
+	rolecredentials "github.com/aws/amazon-ecs-agent/ecs-agent/credentials"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/doctor"
 	"github.com/aws/amazon-ecs-agent/ecs-agent/logger"
+	loggerfield "github.com/aws/amazon-ecs-agent/ecs-agent/logger/field"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/metrics"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/utils/retry"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/utils/ttime"
 	"github.com/aws/amazon-ecs-agent/ecs-agent/wsclient"
+	"github.com/aws/aws-sdk-go/aws/credentials"
 )
 
+const (
+	// heartbeatTimeout is the maximum time to wait between heartbeats
+	// without disconnecting.
+	heartbeatTimeout = 1 * time.Minute
+	heartbeatJitter  = 1 * time.Minute
+
+	// wsRWTimeout is the duration of read and write deadline for the
+	// websocket connection.
+	wsRWTimeout = 2*heartbeatTimeout + heartbeatJitter
+
+	inactiveInstanceReconnectDelay = 1 * time.Hour
+
+	connectionBackoffMin        = 250 * time.Millisecond
+	connectionBackoffMax        = 2 * time.Minute
+	connectionBackoffJitter     = 0.2
+	connectionBackoffMultiplier = 1.5
+
+	// sendCredentialsURLParameterName is the name of the URL parameter
+	// in the ACS URL that is used to indicate if ACS should send
+	// credentials for all tasks on establishing the connection.
+	sendCredentialsURLParameterName = "sendCredentials"
+	inactiveInstanceExceptionPrefix = "InactiveInstanceException"
+
+	// ACS protocol version spec:
+	// 1: default protocol version
+	// 2: ACS will proactively close the connection when heartbeat ACKs are missing
+	acsProtocolVersion = 2
```

Comment:
Just for my understanding. Is this something introduced by ACS and we need to pass in setting up connection (and we just add this to const to avoid magic number in code)? When was it introduced and was there a transition in our system? Is there a consideration that ACS does not use this by default?

Reply:
> Is this something introduced by ACS and we need to pass in setting up connection (and we just add this to const to avoid magic number in code)

Yes, it is introduced by ACS and we need to pass this in when setting up the connection if we want to use protocol version 2. You are correct that the const is for readability purposes and to provide some context with the surrounding comments.

> When was it introduced and was there a transition in our system?

Agent transitioned to use protocol version 2 in Agent version 1.52.0. Here is the corresponding PR: https://github.com/aws/amazon-ecs-agent/pull/2837

Protocol version 2 itself was introduced likely some time in the weeks / months prior to that transition.

> Is there a consideration that ACS does not use this by default?

ACS defaults to protocol version 1 if protocol version is not specified.

---

If you need more information about ACS protocol version, please let me know and I can point you in the right direction!

Reply:
Thanks! This is good enough and is very helpful!